core/src/main/java/cz/o2/proxima/source/BatchSource.java

```diff
@@ -168,6 +164,9 @@ public boolean hasNext() {
 
                 @Override
                 public StreamElement next() {
+                  if (!hasNext()) {
```

Comment:
This doesn't look it will work. I think that the call to `hasNext` will take another element and therefore calling `hasNext()` and then `next` will actually result in one element being skipped. This should probably be fixed either by checking `current` for null, or (better) to actually read the queue in `next` and in `hasNext` just use `peek`.